solana/rpc/providers/ws.py

```diff
@@ -0,0 +1,131 @@
+
+
+class Provider(BaseProvider, FriendlyJsonSerde):
+    """Provider class for websockets."""
+
+    logger = logging.getLogger("solana.providers.WebsocketProvider")
+    _loop = None
+
+    def __init__(
+        self,
+        endpoint_uri: Union[URI, str],
+        websocket_kwargs: Optional[Any] = None,
+        websocket_timeout: int = DEFAULT_WEBSOCKET_TIMEOUT,
+    ) -> None:
+        """Initialize the websocket provider."""
+        self._request_counter = itertools.count()
+        if isinstance(endpoint_uri, str):
+            self.endpoint_uri = URI(endpoint_uri)
+        else:
+            self.endpoint_uri = endpoint_uri
+        self.websocket_timeout = websocket_timeout
+        if self.endpoint_uri is None:
+            self.endpoint_uri = get_default_endpoint()
+        if Provider._loop is None:
+            Provider._loop = _get_threaded_loop()
+        if websocket_kwargs is None:
+            websocket_kwargs = {}
+        else:
+            found_restricted_keys = set(websocket_kwargs.keys()).intersection(RESTRICTED_WEBSOCKET_KWARGS)
+            if found_restricted_keys:
+                raise Exception(
+                    "{0} are not allowed in websocket_kwargs, "
+                    "found: {1}".format(RESTRICTED_WEBSOCKET_KWARGS, found_restricted_keys)
+                )
+        self.conn = PersistentWebSocket(self.endpoint_uri, Provider._loop, websocket_kwargs)
+        super().__init__()
+
+    def __str__(self) -> str:
+        """Return a string representation of the websocket connection."""
+        return "WS connection {0}".format(self.endpoint_uri)
+
+    async def coro_make_request(self, request_data: bytes) -> RPCResponse:
+        """Makes coroutine requests."""
+        async with self.conn as conn:
+            await asyncio.wait_for(conn.send(request_data), timeout=self.websocket_timeout)
+            return json.loads(await asyncio.wait_for(conn.recv(), timeout=self.websocket_timeout))
+
+    def make_request(self, method: RPCMethod, *params: Any) -> RPCResponse:
+        """Make a request to the rpc endpoint."""
+        request_id = next(self._request_counter) + 1
+        self.logger.debug("Making request WebSocket URI: %s, Method: %s, id: %s", self.endpoint_uri, method, request_id)
+        request_data = self.json_encode({"jsonrpc": "2.0", "id": request_id, "method": method, "params": params})
+        if Provider._loop:
+            future = asyncio.run_coroutine_threadsafe(self.coro_make_request(cast(bytes, request_data)), Provider._loop)
+        return future.result()
+
+    def is_connected(self) -> bool:
+        """Health check."""
+        try:
+            if Provider._loop is not None:
+                future = asyncio.run_coroutine_threadsafe(self.coro_make_request(bytes()), Provider._loop)
+                future.result(timeout=self.websocket_timeout)
+                return True
+        except asyncio.TimeoutError:
+            self.logger.debug("Websocket connection timed out.")
```

Comment:
I like how you logged the exception here, can you log exceptions everywhere else too?